CACHE_REQ_SEARCH: Check for filtered users/groups also on cache_req_send()

fidencio opened the pull request:
cache_req_send() may take some shortcuts in case the object is found in
the cache and it's still valid.

This behaviour may lead to exposing filtered users and groups when
they're searched by their uid/gid.

A solution for this issue was proposed on 4ef0b19a but, unfortunately,
didn't take into consideration that this shortcut could be taken.

There are basically two really easy ways to test this issue:
 1) Using enumeration:
    - Set "enumerate = True" in the domain section
    - restart SSSD cleaning up the cache;
    - Wait a little bit till the enumerated users are cached
    - id < uid of a user who is part of the filter_users >

 2) Not using enumeration:
   - getent passwd < uid of a user who is part of the filter_users >
   - Wait a little bit till the user is cached
   - id < same uid used above >

Related:
https://pagure.io/SSSD/sssd/issue/3362

Signed-off-by: Fabiano Fidêncio <fidencio@redhat.com>

fidencio commented:
As this patch should also come with a test, I'm adding the "Changes Requested" label and I will revisit it soon.

fidencio commented:
Just for the record ...

There are two reliable reproducers (at least on my machine) and they are:

Here's the parts of the sssd.conf that are related to this issue:

```
[domain/ipa.example]
...
# Set it accordingly depending on using enumeration or net
enumerate = True
...

[nss]
...
filter_users = user00, user01
filter_groups = user00, user01
entry_negative_timeout = 1
....
```
* When using enumeration:
`# rm -rf /var/log/sssd/sssd* /var/lib/sss/db/* ; systemctl restart sssd; id 1790400001;`
And the answer will be:
`uid=1790400001(user00) gid=1790400001(user00) groups=1790400001(user00)`


* When not using enumeration:
`# rm -rf /var/log/sssd/sssd* /var/lib/sss/db/* ; systemctl restart sssd;  getent passwd 1790400001; sleep 2; id 1790400001`
And the answer will be:
`uid=1790400001(user00) gid=1790400001(user00) groups=1790400001(user00)`

And the expected result for both cases is:
`id: ‘1790400001’: no such user`

pbrezina commented:
I do not like having a temporary memory context inside `cache_req_search_send` and since we are using `cache_req_search_ncache_filter` on two places now, I would like you to change it a little bit so we have clear and unified usage.

The purpose of `cache_req_search_ncache_filter` is to sieve result from cache trough the negative cache and return a new result. The reason why a temporary memory context is needed here is that we return the original result as a "performance improvement" and we do not know in the caller whether we created a new result or reused the original one. We should not use the original result in the caller anymore regardless of what was returned from the filter.

I propose the following change:

```c
static errno_t cache_req_search_ncache_filter(TALLOC_CTX *mem_ctx,
                                              struct cache_req *cr,
                                              struct ldb_result *result,
                                              struct ldb_result **_result)
=>

static errno_t cache_req_search_ncache_filter(TALLOC_CTX *mem_ctx,
                                              struct cache_req *cr,
                                              struct ldb_result **_result)
```

The function will take an input/output argument `_result` and it will free the original result if needed. Then a caller doesn't have to care about it at all.

fidencio commented:
@pbrezina: Thanks a lot for the review.
I'll follow your suggestions and re-submit the patch(es).

I'll do it as soon as I have some tests written for this issue (or if @lslebodn agrees on just having the downstream tests for those).

fidencio commented:
CI: http://sssd-ci.duckdns.org/logs/job/70/89/summary.html

pbrezina commented:
Hmm, also case for midpoint refresh needs to be handled. If `cache_req_expiration_status` returns `CACHE_OBJECT_VALID` or `CACHE_OBJECT_MIDPOINT`, you want to pass the result to the filter and return it immediately (for midpoint case an oob dp request needs to be fired). I think the best solution is to move the filter inside `ret == EOK` section in `done`.

fidencio commented:
@pbrezina: I've updated the last patch according to your suggestion.

Thanks for the review.

pbrezina commented:
Ack.

lslebodn commented:
FYI: much simpler reproducer is
  getent passwd 12785; sleep 2; getent passwd 12785
It works also with "id $num" because it calls getpwuid as well.
Do we want to change commit message?

fidencio commented:
@lslebodn: I've just wrote a simple test for it.

Let me make sure the test actually tests the issue and I'll update the PR in a few.

fidencio commented:
I've updated the commit message and added the test for the change.

I'm removing the "Changes Requested" label for now (as the tests were the missing part).

lslebodn commented on src/tests/intg/test_ldap.py:
```diff
@@ -1076,3 +1076,40 @@ def test_nss_filters(ldap_conn, sanity_nss_filter):
+        pwd.getpwuid(1002)
+    time.sleep(2)
+    with pytest.raises(KeyError):
+        pwd.getpwuid(1002)
```
It would be good to test also getgrgid. and what about merging changes to existing test? 
e.g.
```
diff --git a/src/tests/intg/test_ldap.py b/src/tests/intg/test_ldap.py
index ea7393f5a..077215580 100644
--- a/src/tests/intg/test_ldap.py
+++ b/src/tests/intg/test_ldap.py
@@ -1015,6 +1015,7 @@ def sanity_nss_filter(request, ldap_conn):
             [nss]
             filter_users = user2
             filter_groups = group_two_one_user_groups
+            entry_negative_timeout = 1
         """).format(**locals())
     create_conf_fixture(request, conf)
     create_sssd_fixture(request)
@@ -1067,6 +1068,14 @@ def test_nss_filters(ldap_conn, sanity_nss_filter):
     with pytest.raises(KeyError):
         grp.getgrgid(2019)
 
+    # retest query by id after expiration of negcache -> entry_negative_timeout
+    time.sleep(2)
+
+    with pytest.raises(KeyError):
+        pwd.getpwuid(1002)
+    with pytest.raises(KeyError):
+        grp.getgrgid(2019)
+
     # test non-existing user/group
     with pytest.raises(KeyError):
         pwd.getpwnam("non_existent_user")
```

fidencio replied:
@lslebodn:

I've added the getgrgid() test as well (in the patchset that I'm about to updated) but, unless you really have strong feelings on merging those tests, I'd prefer to keep them separately as different code paths are hit by each test and  I do believe it would be easier to track down the issues in each code path in the way it is.

Again, if you have strong feelings about merging the tests, I'm okay about this as well (although not my preference).

lslebodn replied:
I do not have any preference here.
Thank you for amending test.

ACK to test; and code was already ACKed by Pavel

lslebodn commented:
master:
* 13205258cc17d3833558244251f5adbc98cf34e5
* 4c09cd008967c5c0ec358dc658ffc6fc1cef2697
* c8193b1602cf44740b59f5dfcdc5330508c0c365
